spec/consensus/light-client.md

```diff
@@ -316,3 +316,71 @@ func Backwards(h1,h2) bool {
   return (hash(h2) == old.Header.hash)
  }
 ```
+
+### Failure Detection Module
+
+The light client must be able to detect failures in the trust assumptions.
+Ideally, it should never trust a header, which was forged by malicious
+validators.
+
+### Connectivity requirement
+
+To meet the above requirement, it must be connected to at least one honest
+full node. This is sufficient so long as there's no fork on the main chain.
+
+If there is a fork on the main chain, it means that two full nodes have decided
+on two different headers for the same height. Then the requirement would at
+least require that the lite client is **connected to one honest full node on
+each branch of the fork**.
+
+_Assumption_: no more than `2f` of the nodes are malicious (`f+1` are honest).
+otherwise, they could create more forks.
+
+In practice, this means connecting to 1 or more geographically distributed full
+nodes (called witnesses), which belong to different companies. Note this number
+does not include the primary full node (called primary), which is used for
+obtaining new headers.
+
+_Remark_: we can't guarantee all witnesses won't follow the same branch of a
+fork. to guarantee that, we'll need the light client to be connected to +1/3 of
+nodes, which is impractical.
+
+### Evidence detection
+
+After the light client verifies a new header (`H`) it received from primary, it
+should cross-check `H` with the headers from all witnesses. Cross-checking
+means comparing hashes of the headers. If any two hashes (or more) diverge,
+there's a fork (on the main chain OR phantom fork targeting this light client).
+
+The light client will then need to validate the header it got from a witness
+(`H1`) and verify it's signed by +1/3 of the trusted validator set
+(`CheckSupport(trustedHeader, H1, 1/3)`). If either of these fails, it should
+disconnect from the offending witness.
+
+1. Equivocation
+
+  If some validator double signed, the light client should form & submit a
+  `DuplicateVoteEvidence`.
+
+  After doing so, the light client must stop its operation. The operator will be
+  forced to reset the light client (resetting does not imply deleting the data
+  here) with a new trusted header.
+
+2. Phantom validators
+
+  If there's a vote from a validator outside of the validator set for that
+  height, the light client should reject the header it came from. If it came
+  from primary, it should select a new primary from the list of witnesses and
+  try to continue.
+
+3. Other attacks (Flip-flopping Amnesia & Back to the past, Lunatic)
+
+  Since there is no way for the light client to detect who's lying to it (which
+  full node - primary or one of the witnesses), it must form an evidence and
+  submit it to all connected full nodes (witnesses and primary). The evidence
+  will typically contain a set of diverged headers (including commits for each of
```

Comment:
According to https://github.com/tendermint/tendermint/issues/4182, we should split the evidence into many small InvalidHeaderEvidence, but we don't know which header is invalid, so we'll need to create evidence for both valid and invalid header.

Reply:
Oh, my thinking was that the light client would just send what it saw, and the full node's would handle splitting it into many small pieces since the full nodes will know which is invalid. I think technically the clients could figure it out too with a sequential sync but might as well let the full nodes do it